README.md
# DNA-BOT

**Feb 2022 - DNA-BOT has now been updated to produce scripts that run on the Opentrons OT2 in APIv2.** It also incorporates options for different hardware configurations by producing multiple scripts for each of the 4 steps. The Thermocycler module can now be optionally used in the step 1 clip reactions as well as the step 4 transformations.
In addition, different labware inputs are available through the Graphical User Interface (GUI) or Command Line Interface (CLI).

This work was originally published in [OUP Synthetic Biology](https://academic.oup.com/synbio/article/5/1/ysaa010/5869449)

## Getting Started

Users looking to implement the DNA-BOT workflow are encouraged to consult the [instructions](docs/DNA_BOT_instructions_v1.0.0.pdf). If you are looking to contribute to this project, please raise an issue or pull request. Otherwise, feel free to reach out to [geoffbaldwin](mailto:g.baldwin@imperial.ac.uk).

dnabot can be run in 2 modes:

- with a graphical interface: see `Running the dnabot app` section in [instructions](docs/DNA_BOT_instructions_v1.0.0.pdf). dnabot was developed using Python v3.7. Refer to [requirements.txt](requirements.txt).
- without a graphical interface: you need to specify all settings through the command line, you can see the instructions in the sections below.

## Installation from conda

```bash
conda create --name <myenv>
conda activate <myenv>
conda install -c conda-forge -c brsynth dnabot
```

`<myenv>` has to be replaced by whatever meaningful name that will pleased the user.

## Usage

### Using the GUI

```bash
conda activate <myenv>
python -m dnabot.dnabot_app --help
python -m dnabot.dnabot_app
```

### Using the CLI only

```bash
conda activate <myenv>
python -m dnabot.dnabot_app nogui --help
python -m dnabot.dnabot_app nogui \
    --construct_path /path/to/constructs.csv \
    --source_paths /path/to/linker_parts_coord.csv /path/to/user_parts_coord.csv \
    --output_dir /path/to/output/dir
```

## Command line arguments

### GUI mode

```bash
usage: dnabot_app.py [-h] [--default_settings_file DEFAULT_SETTINGS_FILE] {nogui} ...

DNA assembly using BASIC on OpenTrons.

positional arguments:
  {nogui}               Optional, switch to define settings from the terminal instead of the graphical interface. Type "python dnabot_app.py
                        nogui -h" for more info.

optional arguments:
  -h, --help            show this help message and exit
  --default_settings_file DEFAULT_SETTINGS_FILE
                        Optional, file providing labware IDs and parameter to be used. Default:
                        /Users/tduigou/code/dnabot/dnabot/default_settings.yaml.
```

### No GUI mode

```bash
usage: dnabot_app.py nogui [-h] --construct_path CONSTRUCT_PATH --source_paths SOURCE_PATHS [SOURCE_PATHS ...]
                           [--etoh_well ETOH_WELL] [--soc_column SOC_COLUMN] [--output_dir OUTPUT_DIR]
                           [--template_dir TEMPLATE_DIR]

optional arguments:
  -h, --help            show this help message and exit
  --construct_path CONSTRUCT_PATH
                        File listing constructs to be implemented.
  --source_paths SOURCE_PATHS [SOURCE_PATHS ...]
                        File(s) listing parts to be used in constructs.
  --etoh_well ETOH_WELL
                        Coordinates of the well plate providing ethanol for the purification step. Default: A11
  --soc_column SOC_COLUMN
                        Coordinate of the column plate providing SOC media for the transformation step. Default: 1
  --output_dir OUTPUT_DIR
                        Output directory. Default: same directory than the one containing the 'construct_path' file
  --template_dir TEMPLATE_DIR
                        Template directory. Default: 'template_ot2_scripts' located next to the present script.
```

## Change default values

Use the `--default_settings_file` argument to set different default values. This option is 
available either using the GUI or the CLI interface.

```bash
conda activate <myenv>
python -m dnabot.dnabot_app --default_settings_file /path/to/custom/default_settings.yaml.
```

The default settings file should follow the structure below (yaml file). The 
labware IDs to be used can be updated with the `labwares` section, while the 
parameters for the seperation step are listed in the `parameters` section.

```yaml
labwares:

  # Pipettes #############################################
  # Opentrons P20 Single-Channel Electronic Pipette
  p20_single:
    id: p20_single_gen2

  # Opentrons P300 8-Channel Electronic Pipette
  p300_multi:
    id: p300_multi_gen2

  # Modules ###############################################
  # Opentrons magnetic module (step: purification)
  mag_deck:
    id: magdeck
    # id: magnetic module gen2  # BRS

  # Tip racks #############################################
  # Opentrons 20μL tips rack
  96_tiprack_20ul:
    id: opentrons_96_tiprack_20ul
    # id: tipone_3dprinted_96_tiprack_20ul  # BRS

  # Opentrons 300μL tips rack
  96_tiprack_300ul:
    id: opentrons_96_tiprack_300ul
    # id: tipone_yellow_3dprinted_96_tiprack_300ul  # BRS

  # Plates ################################################
  # Opentrons 4-in-1 tubes rack for 1.5 ml eppendorf tubes (steps: clip, assembly, transformation)
  24_tuberack_1500ul:
    id: e14151500starlab_24_tuberack_1500ul
    # id: opentrons_24_tuberack_eppendorf_1.5ml_safelock_snapcap  # BRS

  # Clip reaction source plate (steps: clip)
  clip_source_plate:
    id: 4ti0960rig_96_wellplate_200ul
    # id: green_96_wellplate_200ul_pcr  # BRS (transparent)

  # Clip reaction plate (steps: clip, purif, assembly)
  clip_plate:
    id: 4ti0960rig_96_wellplate_200ul
    # id: black_96_wellplate_200ul_pcr  # BRS (black)

  # Mix plate (step: purification)
  mix_plate:
    id: 4ti0960rig_96_wellplate_200ul
    # id: black_96_wellplate_200ul_pcr  # BRS (black)
  
  # Final assembly plate (steps: assembly, transformation)
  final_assembly_plate:
    id: 4ti0960rig_96_wellplate_200ul
    # id: black_96_wellplate_200ul_pcr  # BRS (black)
  
  # Transformation plate with thermocycler (step: transformation)
  transfo_plate:
    id: 4ti0960rig_96_wellplate_200ul
    # id: black_96_wellplate_200ul_pcr  # BRS (black)

  # Transformation plate without thermocycler (step: transformation)
  transfo_plate_wo_thermo:
    id: 4ti0960rig_96_wellplate_200ul
    # id: green_96_wellplate_200ul_pcr  # BRS (transparent)

  # Agar plate (transformation step)
  agar_plate:
    id: 4ti0960rig_96_wellplate_200ul
    # id: thermoomnitrayfor96spots_96_wellplate_50ul  # BRS

  # Reservoir plate 21 mL 12 channels (step: purification)
  12_reservoir_21000ul:
    id: 4ti0131_12_reservoir_21000ul
    # id: citadel_12_wellplate_22000ul  # BRS

  # 96 deep well plate 2 mL wells (steps; purification, transormation)
  96_deepwellplate_2ml:
    id: 4ti0136_96_wellplate_2200ul
    # id: transparent_96_wellplate_2ml_deep  # BRS
  
  # Corning 12 Well Plate 6.9 mL Flat (step: transformation)
  12_corning_wellplate:
    id: corning_12_wellplate_6.9ml_flat
    # id: sarstedtcplatte_12_wellplate_6640ul  # BRS
  

parameters:

  # Clip reaction step ####################################
  # Keep the thermocycler lid closed at 4°C at the end of execution?
  # 1 for yes, 0 for no
  clip_keep_thermo_lid_closed:
    value: 0

  # Purification step #####################################
  # Magnetic module height (mm) - purification step
  purif_magdeck_height: 
    value: 20
    # value: 10.8  # BRS

  # Washing time (min) - purification step
  purif_wash_time:
    value: 0.5

  # Bead ratio - purification step
  purif_bead_ratio:
    value: 1.8

  # Incubation time (min) - purification step
  purif_incubation_time:
    value: 5

  # Settling time (min) - purification step
  purif_settling_time:
    value: 2
    # value: 6  # BRS

  # Drying time (min) - purification step
  purif_drying_time:
    value: 5
    # value: 15  # BRS

  # Elution time (min) - purif step
  purif_elution_time:
    value: 2
    # value: 5  # BRS
  
  # Transformation step ###################################
  # Incubation temperature
  transfo_incubation_temp:
    value: 4
    # value: 8  # BRS

  # Incubation time (min)
  transfo_incubation_time:
    value: 20
    # value: 30  # BRS

```

## For developers

### Development installation

After a git clone:

```bash
conda env create -f environment.yaml -n <dev_env>
conda develop -n <dev_env> .
```

You may be prompted to install `conda-build` in your base environment (`conda install conda-build`).
The default conda environment name will be `dnabot-dev` if not specified by `-n <dev-env>`.

Test your installation with:

```bash
conda activate <dev_env>
python -m dnabot.dnabot_app nogui --help
python -m pytest tests
```

To uninstall:

```bash
conda deactivate
conda env remove -n <dev_env>
```

### Tests

You need to install *pytest* if it's not done yet (`conda install pytest`).

```bash
conda install pytest
python -m pytest tests
```

## Authors

The update of DNABOT to APIv2 and improvements to the front end involved the work of several people:

- Thomas Duigou - [tduigou](https://github.com/tduigou)
- Geoff Baldwin - [geoffbaldwin](https://github.com/geoffbaldwin)
- Gizem Buldum  - [gizembuldum](https://github.com/gizembuldum)

Initial work revising the template scripts to run in Opentrons APIv2 was done by a team of Masters students from the 
MRes in Systems and Synthetic Biology at Imperial College London, thanks to:

- Xin Luo
- Ruihan Bai
- Zhenhua Wu
- Lianne Wu
- Ting An Lee
- Xiangming Xu

The original code for DNABOT produced OT2 scripts that ran in APIv1 and was authored by:

- **Matthew C Haines** - [hainesm6](https://github.com/hainesm6)

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

## Acknowledgments

- Marko Storch for all his help with [instructions](docs/DNA_BOT_instructions_v1.0.0.pdf) and DNA-BOT manuscript.
- Geoff Baldwin for all his help with the DNA-BOT manuscript.


CHANGELOG.md
# Changelog

## 3.3.0 (2023-03-01)

- feat: keep the thermo lid closed at 4°C after execution
- feat: refine plate roles and labels

## 3.2.0 (2023-02-17)

- feat: add option to keep thermocycler closed at the end of execution (clip reaction)
- feat: improve GUI layout
- feat: additional output file summarizing labware positions on the deck

## 3.1.0 (2022-02-21)

- fix: transformation with thermocycler template udpate
- fix: add a vertical scrollbar in GUI
- feat: add a transformation template using a 12-well plate
- docs: README update

## 3.0.0 (2021-11-25)

- feat: provide default values from file
- feat: enable execution from CLI only (without graphical interface)
- build: new code architecture for conda recipe
- test: append unit tests
- docs: README update

## v1.0.0 (2019-11-28)

- first released version of DNA-Bot, compatible with API v1 of Opentrons

dnabot/dnabot_gui.py
# -*- coding: utf-8 -*-
"""
Created on Thu May 30 14:35:26 2019

@author: mh2210, tduigou
"""

from __future__ import annotations  # Enable the "hint" feature for objects

import tkinter as tk
from tkinter import ttk
from tkinter import filedialog


def to_numeric_value(str_value: str):
    if float(str_value).is_integer():
        return int(float(str_value))
    else:
        return float(str_value)


class FileSelector:

    def __init__(self, root, irow, title, multiple_files=False):
        # To be used to get the value from outside
        self.output = None
        self.irow = irow
        # Settings
        self.title = title
        self.multiple_files = multiple_files
        # File path as text
        self.entry_text = tk.Entry(root, state="readonly", width=60)
        self.entry_text.insert(0, '')
        # Browse button
        button = tk.Button(root, text='Browse', command=self.browse_file)
        # Position on grid
        button.grid(row=self.irow, columnspan=2)
        self.irow += 1
        self.entry_text.grid(row=self.irow, columnspan=2)
        self.irow += 1

    def browse_file(self):
        if self.multiple_files:
            output = filedialog.askopenfilenames(  # output is a list
                title=self.title,
                filetypes=(('CSV files', '*.csv'), ('All files', '*.*'))
                )
        else:
            # Output is a str
            output = filedialog.askopenfilename(  # output is a str
                title=self.title,
                filetypes=(('CSV files', '*.csv'), ('All files', '*.*'))
                )
        self.output = output
        self.update_text()
    
    def update_text(self):
        self.entry_text.configure(state='normal')
        self.entry_text.delete(0, "end")
        if isinstance(self.output, tuple):
            self.entry_text.insert(0, ', '.join(self.output))
        else:
            self.entry_text.insert(0, self.output)
        self.entry_text.configure(state='readonly')

    def get(self):
        return self.output


class GUI:

    __APP_TITLE = "DNABot App"
    __APP_FONT = ("Helvetica", 12)
    __TROUGH_WELLS = ['A{}'.format(x + 1) for x in range(12)]

    
    def __init__(
        self,
        root: tk.Tk,
        user_settings = dict
        ) -> GUI:
    
        # The set up the GUI backbone
        self.root = root
        self.canvas = tk.Canvas(self.root, width=650, height=840)
        self.frame = tk.Frame(self.canvas)
        self.vsb = tk.Scrollbar(self.root, orient="vertical", command=self.canvas.yview, width=20)
        self.vsb.pack(side="right", fill="y")
        self.canvas.configure(yscrollcommand=self.vsb.set)
        self.canvas.pack(side="left", fill="both", expand=False)
        self.canvas.create_window((0,0), window=self.frame, anchor="nw", tags="self.frame")

        # 
        self.root.title(GUI.__APP_TITLE)
        self.user_settings = user_settings
        self.quit_status = False

        # Intro ===============================================================
        irow = 0
        intro = tk.Message(
            self.frame,
            text=(
                "Welcome to the dnabot App! Please follow these "
                "instructions to create the 4 DNA-BOT scripts:"),
            width=600)
        intro.grid(row=irow, columnspan=2, padx=5, pady=15)

        # Sep =================================================================
        irow += 1
        self.__add_separator(irow)

        # Ethanol & SOC media =================================================
        irow += 1
        message_1 = tk.Message(
            self.frame,
            text=(
                "1 - From the dropdown menus select wells/columns "
                "for ethanol (2_purification script) and SOC media "
                "(4_transformation script)."),
            width=600,
            anchor='w')
        message_1.grid(row=irow, columnspan=2, padx=5, pady=10, sticky='w')

        irow += 1
        etoh_well_label = tk.Label(self.frame, text='Trough well for ethanol during purification:', font=GUI.__APP_FONT)
        etoh_well_label.grid(row=irow, column=0, sticky='e')
        self.etoh_well = tk.StringVar(self.frame)
        self.etoh_well.set(GUI.__TROUGH_WELLS[10])
        etoh_w=tk.OptionMenu(self.frame, self.etoh_well, *tuple(GUI.__TROUGH_WELLS[1:11]))
        etoh_w.grid(row=irow, column=1, sticky=tk.W)
        etoh_w.config(font=GUI.__APP_FONT)

        irow += 1
        soc_column_label = tk.Label(self.frame, text='Deep-well plate column for SOC media during transformation:', font=GUI.__APP_FONT)
        soc_column_label.grid(row=irow, column=0, sticky='e')
        self.soc_column=tk.StringVar(self.frame)
        self.soc_column.set("1")
        soc_w=tk.OptionMenu(self.frame, self.soc_column, *tuple(['{}'.format(x + 1) for x in range(12)]))
        soc_w.grid(row=irow, column=1, sticky='w')
        soc_w.config(font=GUI.__APP_FONT)

        # Sep =================================================================
        irow += 1
        self.__add_separator(irow)

        # Labware IDs =========================================================
        irow += 1
        message_2 = tk.Message(
            self.frame,
            text=(
                "2 - Specify the labware IDs to be used. Leave values "
                "as they are to use the default ones."),
            width=600,
            anchor='w')
        message_2.grid(row=irow, columnspan=2, padx=5, pady=10, sticky='w')

        # Opentrons P20 Single-Channel Electronic Pipette
        irow += 1
        self.labware_p10_single_entry = self.__make_labware_entry(
            label="Opentrons P20 Single-Channel Electronic Pipette",
            labware_id='p20_single',
            irow=irow)
        # Opentrons P300 8-Channel Electronic Pipette
        irow += 1
        self.labware_p300_multi_entry = self.__make_labware_entry(
            label="Opentrons P300 8-Channel Electronic Pipette",
            labware_id='p300_multi',
            irow=irow)
        # Opentrons magnetic module
        irow += 1
        self.labware_mag_deck_entry = self.__make_labware_entry(
            label="Opentrons magnetic module",
            labware_id='mag_deck',
            irow=irow)
        # Opentrons 4-in-1 tubes rack for 1.5 ml eppendorf tubes
        irow += 1
        self.labware_24_tuberack_1500ul_entry = self.__make_labware_entry(
            label="Opentrons 4-in-1 tubes rack",
            labware_id='24_tuberack_1500ul',
            irow=irow)
        # Opentrons 10μL tips rack
        irow += 1
        self.labware_96_tiprack_20ul_entry = self.__make_labware_entry(
            label="Opentrons 20μL tips rack",
            labware_id='96_tiprack_20ul',
            irow=irow)
        # Opentrons 300μL tips rack
        irow += 1
        self.labware_96_tiprack_300ul_entry = self.__make_labware_entry(
            label="Opentrons 300μL tips rack",
            labware_id='96_tiprack_300ul',
            irow=irow)
        # Clip reaction source plate (steps: clip)
        irow += 1
        self.labware_clip_source_plate_entry = self.__make_labware_entry(
            label="Clip reaction source plate (steps: clip)",
            labware_id='clip_source_plate',
            irow=irow)
        # Clip reaction plate (steps: clip, purif, assembly)
        irow += 1
        self.labware_clip_plate_entry = self.__make_labware_entry(
            label="Clip reaction plate (steps: clip, purification, assembly)",
            labware_id='clip_plate',
            irow=irow)
        # Mix plate (step: purification)
        irow += 1
        self.labware_mix_plate_entry = self.__make_labware_entry(
            label="Mix plate (step purification)",
            labware_id='mix_plate',
            irow=irow)
        # Final assembly plate (steps: assembly, transformation)
        irow += 1
        self.labware_final_assembly_plate_entry = self.__make_labware_entry(
            label="Final assembly plate (steps: assembly, transformation)",
            labware_id='final_assembly_plate',
            irow=irow)
        # Transformation plate (step: transformation)
        irow += 1
        self.labware_transfo_plate_entry = self.__make_labware_entry(
            label="Transformation plate (step: transformation)",
            labware_id='transfo_plate',
            irow=irow)
        # Transformation plate without thermocycler (step: transformation)
        irow += 1
        self.labware_transfo_plate_wo_thermo_entry = self.__make_labware_entry(
            label="Transformation plate without thermocycler (step: transformation)",
            labware_id='transfo_plate_wo_thermo',
            irow=irow)
        # Agar plate (transformation step)
        irow += 1
        self.agar_plate_entry = self.__make_labware_entry(
            label="Agar plate (transformation step)",
            labware_id='agar_plate',
            irow=irow)
        # Reservoir plate 21 mL 12 channels
        irow += 1
        self.labware_12_reservoir_21000ul_entry = self.__make_labware_entry(
            label="Reservoir plate 21 mL 12 channels",
            labware_id='12_reservoir_21000ul',
            irow=irow)
        # 96 deep well plate 2 mL wells
        irow += 1
        self.labware_96_deepwellplate_2ml_entry = self.__make_labware_entry(
            label="96 deep well plate 2 mL wells",
            labware_id='96_deepwellplate_2ml',
            irow=irow)
        # Corning 12 Well Plate 6.9 mL Flat
        irow += 1
        self.labware_12_corning_wellplate_entry = self.__make_labware_entry(
            label="Corning 12 Well Plate 6.9 mL Flat",
            labware_id="12_corning_wellplate",
            irow=irow)
        
        # Sep =================================================================
        irow += 1
        self.__add_separator(irow)

        # Parameters for the clip reaction step ===============================
        irow += 1
        message_3 = tk.Message(
            self.frame,
            text="3 - Specify parameters for the clip reaction step.",
            width=600,
            anchor='w')
        message_3.grid(row=irow, columnspan=2, padx=5, pady=10, sticky='w')
        irow += 1
        self.param_clip_thermo_lid_closed = self.__make_parameter_entry(
            label="Keep the thermocycler lid closed at 4°C at the end of execution? \n (1 for yes, 0 for no)",
            parameter_id="clip_keep_thermo_lid_closed",
            irow=irow)

        # Sep =================================================================
        irow += 1
        self.__add_separator(irow)

        # Parameters for the purification step ================================
        irow += 1
        message_4 = tk.Message(
            self.frame,
            text="4 - Specify parameters for the purification step.",
            width=600,
            anchor='w')
        message_4.grid(row=irow, columnspan=2, padx=5, pady=10, sticky='w')
        irow += 1
        self.param_purif_magdeck_height = self.__make_parameter_entry(
            label="Magnetic module height (mm)",
            parameter_id="purif_magdeck_height",
            irow=irow)
        irow += 1
        self.param_purif_wash_time = self.__make_parameter_entry(
            label="Washing time (min)",
            parameter_id="purif_wash_time",
            irow=irow)
        irow += 1
        self.param_purif_bead_ratio = self.__make_parameter_entry(
            label="Bead ratio",
            parameter_id="purif_bead_ratio",
            irow=irow)
        irow += 1
        self.param_purif_incubation_time = self.__make_parameter_entry(
            label="Incubation time (min)",
            parameter_id="purif_incubation_time",
            irow=irow)
        irow += 1
        self.param_purif_settling_time = self.__make_parameter_entry(
            label="Settling time (min)",
            parameter_id="purif_settling_time",
            irow=irow)
        irow += 1
        self.param_purif_drying_time = self.__make_parameter_entry(
            label="Drying time (min)",
            parameter_id="purif_drying_time",
            irow=irow)
        irow += 1
        self.param_purif_elution_time = self.__make_parameter_entry(
            label="Elution time (min)",
            parameter_id="purif_elution_time",
            irow=irow)

        # Sep =================================================================
        irow += 1
        self.__add_separator(irow)

        # Parameters for the transformation step ==============================
        irow += 1
        message_5 = tk.Message(
            self.frame,
            text="5 - Specify parameters for the transformation step.",
            width=600,
            anchor='w')
        message_5.grid(row=irow, columnspan=2, padx=5, pady=10, sticky='w')
        irow += 1
        self.param_transfo_incubation_temp = self.__make_parameter_entry(
            label="Incubation temperature (°C)",
            parameter_id="transfo_incubation_temp",
            irow=irow)
        irow += 1
        self.param_transfo_incubation_time = self.__make_parameter_entry(
            label="Incubation time (min)",
            parameter_id="transfo_incubation_time",
            irow=irow)

        # Sep =================================================================
        irow += 1
        self.__add_separator(irow)

        # Construct CSV file ==================================================
        irow += 1
        message_6 = tk.Message(
            self.frame,
            text="6 - Select the CSV file describing constructs.",
            width=600,
            anchor='w')
        message_6.grid(row=irow, columnspan=2, padx=5, pady=10, sticky='w')
        irow += 1
        self.construct_file_selector = FileSelector(self.frame, irow, title='Construct CSV file', multiple_files=False)
        irow = self.construct_file_selector.irow

        # Sep =================================================================
        irow += 1
        self.__add_separator(irow)

        # Source CSV files ====================================================
        irow += 1
        message_7 = tk.Message(
            self.frame,
            text=(
                "7 - Select up to 6 csv files describing plates "
                "containing BASIC parts and linkers. If all files "
                "are not within one folder, absolute paths should "
                "be given."),
            width=600,
            anchor='w')
        message_7.grid(row=irow, columnspan=2, padx=5, pady=10, sticky='w')
        irow += 1
        self.source_files_selector = FileSelector(self.frame, irow, title='Source CSV files', multiple_files=True)
        irow = self.source_files_selector.irow

        # Sep =================================================================
        irow += 1
        self.__add_separator(irow)

        # # White space
        # irow += 1
        # spacer = tk.Label(self.frame, text="", font=GUI.__APP_FONT)
        # spacer.grid(row=irow, columnspan=2, padx=5, pady=10)

        # Quit and generate buttons
        irow += 1
        quit_button = tk.Button(self.frame, text='QUIT', fg='red', command=self.quit, font=GUI.__APP_FONT)
        quit_button.grid(row=irow, column=0, pady=10)
        generate_button=tk.Button(self.frame, text='GENERATE', command=self.generate, font=GUI.__APP_FONT)
        generate_button.grid(row=irow, column=1, pady=10)

        # Upate scroll region info
        self.root.update()  # Required to refresh canvas scrollreion
        self.canvas.configure(scrollregion = self.canvas.bbox("all"))

        # 
        self.frame.mainloop()


    def quit(self):
        self.quit_status=True
        self.root.quit()

    def generate(self):
        # Step 1
        self.user_settings['etoh_well'] = self.etoh_well.get()
        self.user_settings['soc_column'] = self.soc_column.get()

        # Labware IDs
        self.user_settings['labwares']['p20_single']['id'] = self.labware_p10_single_entry.get()
        self.user_settings['labwares']['p300_multi']['id'] = self.labware_p300_multi_entry.get()
        self.user_settings['labwares']['mag_deck']['id'] = self.labware_mag_deck_entry.get()
        self.user_settings['labwares']['24_tuberack_1500ul']['id'] = self.labware_24_tuberack_1500ul_entry.get()
        self.user_settings['labwares']['96_tiprack_20ul']['id'] = self.labware_96_tiprack_20ul_entry.get()
        self.user_settings['labwares']['96_tiprack_300ul']['id'] = self.labware_96_tiprack_300ul_entry.get()
        self.user_settings['labwares']['clip_source_plate']['id'] = self.labware_clip_source_plate_entry.get()
        self.user_settings['labwares']['clip_plate']['id'] = self.labware_clip_plate_entry.get()
        self.user_settings['labwares']['mix_plate']['id'] = self.labware_mix_plate_entry.get()
        self.user_settings['labwares']['final_assembly_plate']['id'] = self.labware_final_assembly_plate_entry.get()
        self.user_settings['labwares']['transfo_plate']['id'] = self.labware_transfo_plate_entry.get()
        self.user_settings['labwares']['transfo_plate_wo_thermo']['id'] = self.labware_transfo_plate_wo_thermo_entry.get()
        self.user_settings['labwares']['agar_plate']['id'] = self.agar_plate_entry.get()
        self.user_settings['labwares']['12_reservoir_21000ul']['id'] = self.labware_12_reservoir_21000ul_entry.get()
        self.user_settings['labwares']['96_deepwellplate_2ml']['id'] = self.labware_96_deepwellplate_2ml_entry.get()
        self.user_settings['labwares']['12_corning_wellplate']['id'] = self.labware_12_corning_wellplate_entry.get()
        # Parameters for the clip reaction step
        self.user_settings["parameters"]["clip_keep_thermo_lid_closed"]["value"] = to_numeric_value(self.param_clip_thermo_lid_closed.get())
        # Parameters for the purification step
        self.user_settings['parameters']['purif_magdeck_height']['value'] = to_numeric_value(self.param_purif_magdeck_height.get())
        self.user_settings['parameters']['purif_wash_time']['value'] = to_numeric_value(self.param_purif_wash_time.get())
        self.user_settings['parameters']['purif_bead_ratio']['value'] = to_numeric_value(self.param_purif_bead_ratio.get())
        self.user_settings['parameters']['purif_incubation_time']['value'] = to_numeric_value(self.param_purif_incubation_time.get())
        self.user_settings['parameters']['purif_settling_time']['value'] = to_numeric_value(self.param_purif_settling_time.get())
        self.user_settings['parameters']['purif_drying_time']['value'] = to_numeric_value(self.param_purif_drying_time.get())
        self.user_settings['parameters']['purif_elution_time']['value'] = to_numeric_value(self.param_purif_elution_time.get())
        # Parameters for the transformation step
        self.user_settings['parameters']['transfo_incubation_temp']['value'] = to_numeric_value(self.param_transfo_incubation_temp.get())
        self.user_settings['parameters']['transfo_incubation_time']['value'] = to_numeric_value(self.param_transfo_incubation_time.get())
        # Construct CSV file
        self.user_settings['construct_path'] = self.construct_file_selector.get()
        # Source CSV files
        self.user_settings['sources_paths'] = self.source_files_selector.get()
        self.root.quit()

    def __make_labware_entry(self, label, labware_id, irow):
        labware_label = tk.Label(self.frame, text=label, font=GUI.__APP_FONT)
        labware_label.grid(row=irow, column=0, sticky='e')
        labware_entry = tk.Entry(self.frame, width=30)
        labware_entry.insert(0, self.user_settings["labwares"][labware_id]['id'])
        labware_entry.grid(row=irow, column=1, sticky='w')
        return labware_entry

    def __make_parameter_entry(self, label, parameter_id, irow, parameter_value="value"):
        parameter_label = tk.Label(self.frame, text=label, font=GUI.__APP_FONT)
        parameter_label.grid(row=irow, column=0, sticky='e')
        parameter_entry = tk.Entry(self.frame, width=30)
        parameter_entry.insert(0, self.user_settings["parameters"][parameter_id][parameter_value])
        parameter_entry.grid(row=irow, column=1, sticky='w')
        return parameter_entry

    def __add_separator(self, irow):
        ttk.Separator(
            self.frame,
            orient=tk.HORIZONTAL
        ).grid(
            row=irow,
            columnspan=2,
            sticky="ew"
        )